### PR TITLE
Limited Quantity of a Block

### PR DESCRIPTION
**Pull request by Hamms**
[Spec (also WIP)](https://docs.google.com/document/d/1yzj8SFoovDr57vrB7_4hJPYswzL4jFb19P0xlP_pLT8/edit)

Adds the option for levelbuilders to specify a limit for blocks in the
toolbox. Such blocks will only allow the specified number of blocks to
be added to the workspace, and will become disabled once their limit is
reached.

image

TODOs remaining:
- [x] tests
- [x] copy/paste protection

post-publish TODOs:
- [ ] create an allthethings level using this, create eyes and UI tests pointing at that level

### CODE REVIEW

File: blockly-core/core/ui/block.js [hunk 1]
```diff
@@ -17,6 +17,8 @@
  * limitations under the License.
  */
 
+/* global Blockly, goog */
```
Comment (bcjordan) [inline blockly-core/core/ui/block.js:20]: Is this needed? Would expect the `goog.require`s below to cover it

Comment (Hamms) [inline blockly-core/core/ui/block.js:20]: Not for eslint :P

File: blockly-core/core/ui/block.js [hunk 2]
```diff
@@ -1454,6 +1472,69 @@ Blockly.Block.prototype.setMovable = function(movable) {
 };
 
 /**
+ * Get whether this block represents a toolbox block that can only be
+ * used a limited number of times
+ * @return {boolean}
+ */
+Blockly.Block.prototype.hasLimit = function() {
+  return (this.isInFlyout || Blockly.editBlocks) && this.limit_ !== undefined;
```
Comment (bcjordan) [inline blockly-core/core/ui/block.js]: What does the `Blockly.editBlocks` part of this do? Would there be a way to name & extract the `(this.isInFlyout || Blockly.editBlocks)` clause of this? e.g. `var shouldDisplayLimits = this.isInFlyout || Blockly.editBlocks;` if that's what it means

File: blockly-core/core/ui/block.js [hunk 3]
```diff
@@ -1454,6 +1472,69 @@ Blockly.Block.prototype.setMovable = function(movable) {
...
+ * @param {number} limit
+ */
+Blockly.Block.prototype.setLimit = function(limit) {
+  limit = parseInt(limit);
+  if (isNaN(limit)) {
+    limit = undefined;
+  }
+  // only allow limits for toolbox blocks or when editing blocks (so
+  // that we can edit toolbox blocks)
+  if (this.isInFlyout || Blockly.editBlocks) {
```
Comment (bcjordan) [inline blockly-core/core/ui/block.js]: Ah makes sense now. Could maybe extract both usages of this into a `block.shouldRenderLimit()` helper instead, and move the above comment into there

File: blockly-core/core/ui/block_space/flyout.js [hunk 4]
```diff
@@ -740,6 +769,48 @@ Blockly.Flyout.prototype.filterForCapacity_ = function() {
 };
 
 /**
+ * Update the block counts for limited-quantity workspace blocks
+ * @private
+ */
+Blockly.Flyout.prototype.updateBlockLimits_ = function() {
+  Object.keys(this.blockLimits_)
+      .forEach(function (type) {
```
Comment (bcjordan) [inline blockly-core/core/ui/block_space/flyout.js]: Another option is [goog.object.forEach](http://google.github.io/closure-library/api/namespace_goog_object.html#forEach), e.g. `goog.object.forEach(this.blockLimits, function(value, type, object) { object.resetTotal() })` (not sure why it's value, then key... 😆 )

**Comment by bcjordan** [comment]
Had call with @Hamms, discussed encapsulating some of the count tracking into a `BlockLimit` component on the flyout which emits updates on changes.

**Comment by Hamms** [comment]
I've moved the functionality into a `BlockLimit` component. We had discussed the possibility of structuring the component to emit events on update and to have `Flyout` bind those events to the display update, but after playing around with a few implementations of that, I think I like this approach better. LMK what you think!

File: blockly-core/core/ui/block.js (hunk 1 above)
Comment (islemaster) [inline blockly-core/core/ui/block.js:20]: I didn't even realize we were running eslint against blockly-core.  Seems like it should have its own linting solution.

Comment (Hamms) [inline blockly-core/core/ui/block.js:20]: Oh, hmm. That might just be a product of my own development environment; looks like we don't lint blockly-core at all.

**Comment by islemaster** [comment]
This looks fantastic.  My one question is, should there be a test to ensure that block limit bubbles are _not_ shown when no blocks have limits?

**Comment by Hamms** [comment]
I do plan to create a UI/Eyes test once this is deployed to levelbuilder, if that's what you mean.

File: blockly-core/core/utils/xml.js [hunk 5]
```diff
@@ -362,6 +369,11 @@ Blockly.Xml.domToBlock = function(blockSpace, xmlBlock) {
   if (userCreated) {
     block.userCreated = (userCreated === 'true');
   }
+  var limit = xmlBlock.getAttribute('limit');
+  var shouldShowLimits = blockSpace.isFlyout || Blockly.editBlocks;
```
Comment (islemaster) [inline blockly-core/core/utils/xml.js]: Should this be an xor [as expressed here](https://github.com/code-dot-org/code-dot-org/pull/8751/files#diff-19bada970605b4e1a316b8460d76be00R203)?

Comment (Hamms) [inline blockly-core/core/utils/xml.js]: yep!

**Pull request merged**